### PR TITLE
Improve Outbound Adapter

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
## Purpose
Following improvements are done to the outbound adapter.
- Upgrade adapter with Apache HTTP Core 5 and Client 5 (Inorder to resolve MTLS issue with connection pool).
- Add MTLS authentication support for event publishing and event configuration (Initially proposed via https://github.com/wso2-extensions/identity-outbound-adapters/pull/21)
- Implement connection pool for event configuration client with MTLS.
- Add server config support for disabling topic deletion.

## Related PRs
- https://github.com/wso2/orbit/pull/974

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/config/WebSubAdapterConfiguration.java [hunk 1]
```diff
@@ -33,7 +33,9 @@ public class WebSubAdapterConfiguration {
 
     private static final String ADAPTER_ENABLED_CONFIG = "adapter.websubhub.enabled";
     private static final String ENCRYPTION_ENABLED_CONFIG = "adapter.websubhub.encryptionEnabled";
+    private static final String TOPIC_DELETION_DISABLED_CONFIG = "adapter.websubhub.topicDeletionDisabled";
```
Comment (ashendes) [inline components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/config/WebSubAdapterConfiguration.java:36]: Pls link all issues of the feature developments included in this PR.

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java [hunk 2]
```diff
@@ -51,86 +67,149 @@ public class ClientManager {
...
         } catch (IOException e) {
             throw handleServerException(ERROR_CREATING_ASYNC_HTTP_CLIENT, e);
         }
 
-        RequestConfig config = createRequestConfig();
-        HttpAsyncClientBuilder httpClientBuilder = HttpAsyncClients.custom().setDefaultRequestConfig(config);
-        addSslContext(httpClientBuilder);
-        httpClientBuilder.setConnectionManager(connectionManager);
-        httpAsyncClient = httpClientBuilder.build();
+        LOG.info("Creating HTTP async client manager for WebSub outbound adapter.");
```
Comment (ashendes) [inline components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java:88]: Lets make these debug logs. Also use the name `Websubhub`

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java [hunk 3]
```diff
@@ -51,86 +67,149 @@ public class ClientManager {
...
+        httpAsyncClient = HttpAsyncClients.custom()
+                .setConnectionManager(asyncConnectionManager)
+                .setIOReactorConfig(IOReactorConfig.custom().build())
+                .setDefaultRequestConfig(RequestConfig.custom().build())
+                .build();
+
         httpAsyncClient.start();
+
+        // Create the http client.
+        LOG.info("Creating HTTP client manager for WebSub outbound adapter.");
```
Comment (ashendes) [inline components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java:98]: Debug log.

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/config/WebSubAdapterConfiguration.java (hunk 1 above)
Comment (ThaminduDilshan) [reply components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/config/WebSubAdapterConfiguration.java:36]: Linked

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java (hunk 2 above)
Comment (ThaminduDilshan) [reply components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java]: Resolved

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java (hunk 3 above)
Comment (ThaminduDilshan) [reply components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java]: Resolved

**Review by sumedhe — APPROVED** [review]

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java [hunk 4]
```diff
@@ -51,86 +67,155 @@ public class ClientManager {
...
+                .setMaxConnPerRoute(maxConnectionsPerRoute)
+                .setDefaultConnectionConfig(createConnectionConfig())
+                .setDefaultTlsConfig(createTlsConfig())
+                .build();
+    }
+
+    private SSLContext getSSLContext() throws WebSubAdapterServerException {
+
+        if (sslContext == null) {
+            LOG.info("Creating SSL context for WebSub outbound adapter.");
```
Comment (ashendes) [inline components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java:187]: This log would seem out of place as the others are debug logs.

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java [hunk 5]
```diff
@@ -191,60 +192,51 @@ public static void makeAsyncAPICall(SecurityEventTokenPayload securityEventToken
...
-                String responsePhrase = response.getStatusLine().getReasonPhrase();
+                int responseCode = response.getCode();
+                String responsePhrase = response.getReasonPhrase();
                 if (log.isDebugEnabled()) {
                     log.debug("WebSubHub request completed. Response code: " + responseCode);
                 }
-                handleResponseCorrelationLog(request, requestStartTime, RequestStatus.COMPLETED.getStatus(),
-                        String.valueOf(responseCode), responsePhrase);
 
+                 handleResponseCorrelationLog(request, requestStartTime, RequestStatus.COMPLETED.getStatus(),
```
Comment (ashendes) [inline components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java:209]: Incorrect indentation?

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java [hunk 6]
```diff
@@ -191,60 +192,51 @@ public static void makeAsyncAPICall(SecurityEventTokenPayload securityEventToken
...
                 }
             }
 
             @Override
-            public void failed(final Exception ex) {
+            public void failed(Exception ex) {
 
-                handleResponseCorrelationLog(request, requestStartTime, RequestStatus.FAILED.getStatus(),
-                        ex.getMessage());
+                 handleResponseCorrelationLog(request, requestStartTime, RequestStatus.FAILED.getStatus(),
```
Comment (ashendes) [inline components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java:228]: Indentation error

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java [hunk 7]
```diff
@@ -191,60 +192,51 @@ public static void makeAsyncAPICall(SecurityEventTokenPayload securityEventToken
...
+                         ex.getMessage());
                 log.error("Publishing event data to WebSubHub failed. ", ex);
             }
 
             @Override
             public void cancelled() {
 
-                handleResponseCorrelationLog(request, requestStartTime, RequestStatus.CANCELLED.getStatus());
+                 handleResponseCorrelationLog(request, requestStartTime, RequestStatus.CANCELLED.getStatus());
```
Comment (ashendes) [inline components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java:236]: Indentation error

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java (hunk 4 above)
Comment (ThaminduDilshan) [reply components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/internal/ClientManager.java]: Resolved

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java (hunk 5 above)
Comment (ThaminduDilshan) [reply components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java]: Resolved

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java (hunk 6 above)
Comment (ThaminduDilshan) [reply components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java]: Resolved

File: components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java (hunk 7 above)
Comment (ThaminduDilshan) [reply components/org.wso2.identity.outbound.adapter.websubhub/src/main/java/org/wso2/identity/outbound/adapter/websubhub/util/WebSubHubAdapterUtil.java]: Resolved

**Review by ashendes — APPROVED** [review]

**Pull request merged**